### PR TITLE
[SYCL] Make SYCL math functions overloads instead of templates

### PR DESCRIPTION
**Pull request by KornevNikita**
It was decided to change all built-in functions from templates to overloads. This patch changes non-marray part of "4.17.5. Math functions".

Spec: https://github.com/KhronosGroup/SYCL-Docs/pull/428

### CODE REVIEW

**Comment by dm-vodopyanov** [comment]
Has math_builtin_api SYCL CTS test some regressions or not with the patch?

**Comment by bader** [comment]
> It was decided to change all built-in functions from templates to overloads.

Why?

**Comment by dm-vodopyanov** [comment]
> > It was decided to change all built-in functions from templates to overloads.
> 
> Why?

Replied internally.

File: sycl/include/sycl/builtins.hpp [hunk 1]
```diff
@@ -48,18 +48,299 @@ template <class T, int N> marray<T, N> to_marray(vec<T, N> x) {
 namespace __sycl_std = __host_std;
 #endif
 
+#define COMMA ,
+
+#define DEF_BUILTIN_VEC(TYPE)                                                  \
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/builtins.hpp:53]: Oh, one more thing - we should start all internal macro names with a prefix `__SYCL_` to distinguish them from user-level macros.

Comment (steffenlarsen) [reply sycl/include/sycl/builtins.hpp:53]: I agree, we should probably follow that naming. In addition, we should probably also undefine them at the end of the header, since they are just intended as generators.

File: sycl/include/sycl/builtins.hpp [hunk 2]
```diff
@@ -48,18 +48,299 @@ template <class T, int N> marray<T, N> to_marray(vec<T, N> x) {
...
+  DEF_BUILTIN_GENFLOATF                                                        \
+  DEF_BUILTIN_GENFLOATD                                                        \
+  DEF_BUILTIN_GENFLOATH
+
+#define DEF_BUILTIN_GENGEOFLOAT                                                \
+  DEF_BUILTIN_GENGEOFLOATF                                                     \
+  DEF_BUILTIN_GENGEOFLOATD                                                     \
+  DEF_BUILTIN_GENGEOFLOATH
+
+// TODO: Replace with overloads.
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/include/sycl/builtins.hpp:317]: I guess it's marray related logic, as it used only [here](https://github.com/intel/llvm/pull/9753/files#diff-6926fd84a1f072d66519ec548ce96a37092088488d3d51014040ff3b749cf7e1R406) and [here](https://github.com/intel/llvm/pull/9753/files#diff-6926fd84a1f072d66519ec548ce96a37092088488d3d51014040ff3b749cf7e1R480)

**Comment by KornevNikita** [comment]
@steffenlarsen @dm-vodopyanov this patch finally passes CI, but I'm going to reduce it a bit by outlining the defines - https://github.com/intel/llvm/pull/10026.

File: sycl/include/sycl/builtins.hpp (hunk 1 above)
Comment (KornevNikita) [reply sycl/include/sycl/builtins.hpp]: The defines are here now https://github.com/intel/llvm/pull/10026

File: sycl/include/sycl/builtins.hpp [hunk 3]
```diff
@@ -602,207 +599,261 @@ __SYCL_MATH_FUNCTION_3_OVERLOAD(mad) __SYCL_MATH_FUNCTION_3_OVERLOAD(mix)
...
-fract(T x, T2 iptr) __NOEXC {
+// genfloat fmod (genfloat x, genfloat y)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE fmod(TYPE x, TYPE y) __NOEXC {                                   \
+    return __sycl_std::__invoke_fmod<TYPE>(x, y);                              \
+  }
+    __SYCL_DEF_BUILTIN_GENFLOAT
+#undef __SYCL_BUILTIN_DEF
+
+    // svgenfloat fract (svgenfloat x, genfloatptr iptr)
```
Comment (KornevNikita) [inline sycl/include/sycl/builtins.hpp:852]: I just noticed there are no some built-ins that accept arguments with gen*ptr:

> genfloat fract(genfloat x, genfloatptr iptr)
> genfloat frexp(genfloat x, genintptr exp)
> genfloat lgamma_r(genfloat x, genintptr signp)
> genfloat modf(genfloat x, genfloatptr iptr)
> genfloat remquo(genfloat x, genfloat y, genintptr quo)
> genfloat sincos(genfloat x, genfloatptr cosval)

I think it will be better to create a separate patch for them as this patch is already too big for review.

File: sycl/include/sycl/builtins.hpp [hunk 4]
```diff
@@ -48,18 +48,299 @@ template <class T, int N> marray<T, N> to_marray(vec<T, N> x) {
 namespace __sycl_std = __host_std;
 #endif
 
+#define COMMA ,
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/include/sycl/builtins.hpp]: There are no new regressions with this patch

File: sycl/include/sycl/builtins.hpp [hunk 5]
```diff
@@ -2514,89 +2614,159 @@ __SYCL_NATIVE_MATH_FUNCTION_2_OVERLOAD(powr)
...
+__SYCL_DEF_BUILTIN_FLOAT_SCALAR
+__SYCL_DEF_BUILTIN_FLOAT_VEC
+#undef __SYCL_BUILTIN_DEF
+
+// genfloatf recip (genfloatf x)
+#define __SYCL_BUILTIN_DEF(TYPE)                                               \
+  inline TYPE recip(TYPE x) __NOEXC {                                          \
+    return __sycl_std::__invoke_native_recip<TYPE>(x);                         \
+  }
+// TODO: Replace with __SYCL_DEF_BUILTIN_GENFLOAT when merged with above
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/builtins.hpp:2721]: The macro is in intel/llvm already, maybe update here?

Comment (KornevNikita) [reply sycl/include/sycl/builtins.hpp]: 49ec80d

**Review by dm-vodopyanov — APPROVED** [review]

**Comment by KornevNikita** [comment]
@intel/dpcpp-esimd-reviewers could you please take a look at esimd-related part?

**Comment by dm-vodopyanov** [comment]
@intel/dpcpp-esimd-reviewers, friendly ping.

File: sycl/test-e2e/ESIMD/kmeans/kmeans.cpp [hunk 6]
```diff
@@ -75,10 +75,10 @@ bool verify_result(Centroid4 *centroids4, // gpu centroids result
   int k = 0;
   int j = 0;
   for (auto i = 0; i < NUM_CENTROIDS_ACTUAL; i++) {
-    float errX = fabs(centroids4[j].x[k] - centroids[i].x) /
-                 max(fabs(centroids4[j].x[k]), fabs(centroids[i].x));
-    float errY = fabs(centroids4[j].y[k] - centroids[i].y) /
-                 max(fabs(centroids4[j].y[k]), fabs(centroids[i].y));
+    float errX = std::fabs(centroids4[j].x[k] - centroids[i].x) /
```
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/test-e2e/ESIMD/kmeans/kmeans.cpp:78]: Yes, this may affect users that have `using namespace sycl;` as the cmath builtins would have been taking precedence while the SYCL builtins weren't overloads. Arguably, this might also help users find bugs in their code as using cmath builtins in a SYCL kernel isn't guaranteed too work.

[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/test-e2e/ESIMD/kmeans/kmeans.cpp:78]: ok thanks, seems legit to me and sounds like you guys thought about the impact

**Review by sarnex — APPROVED** [review]
esimd changes look good to me, slava may have more feedback when he is back from vacation but i don't see anything insane
